// react-charts/.history/src/utils/Context_20180602172034.js
import React from 'react'
import hoistNonReactStatics from 'hoist-non-react-statics'

const Context = React.createContext({})
const HoverContext = React.createContext({})

export default Context

class SafeRender extends React.Component {
  shouldComponentUpdate (nextProps) {
    if (nextProps.filter) {
      return nextProps.filter(this.state, nextProps.state, meta)
    }
    return true
  }
  render () {
    return children
  }
}

export const Connect = Consumer => (subscribe, options) => Comp => {
  const subscriber = typeof subscribe({}) === 'function' ? subscribe() : subscribe
  class Connected extends React.Component {
    render () {
      filter: (oldState, newState, meta) => meta.type !== 'pointer',
      return (
        <Consumer>
          {({ dispatch, _meta, ...rest }) => (
            <Comp {...subscriber(rest, this.props)} {...this.props} dispatch={dispatch} />
          )}
        </Consumer>
      )
    }
  }
  hoistNonReactStatics(Connected, Comp)
  return Connected
}
