// react-charts/.history/src/utils/Selectors_20180602092338.js
import Memoize from './Memoize'

export default {
  primaryAxes: () =>
    Memoize(
      state => [state.axes],
      axes => {
        if (!axes) return []
        return Object.keys(axes)
          .map(key => {
            if (axes[key].primary) {
              return axes[key]
            }
            return null
          })
          .filter(Boolean)
      }
    ),

  secondaryAxes: () =>
    Memoize(
      state => [state.axes],
      axes => {
        if (!axes) return []
        return Object.keys(axes)
          .map(key => {
            if (!axes[key].primary) {
              return axes[key]
            }
            return null
          })
          .filter(Boolean)
      }
    ),

  offset: () =>
    Memoize(
      state => [
        (state && state.offset && state.offset.left) || 0,
        (state && state.offset && state.offset.top) || 0,
      ],
      (left, top) => ({
        left,
        top,
      })
    ),

  gridX: () =>
    Memoize(
      state => [(state && state.padding && state.padding.left) || 0, state && state.axesDimensions],
      (paddingLeft, axesDimensions) => {
        const axesLeftWidth =
          (axesDimensions && axesDimensions.left && sumObjBy(axesDimensions.left, 'width')) || 0
        const axesTopLeft =
          (axesDimensions && axesDimensions.top && sumObjBy(axesDimensions.top, 'left')) || 0
        const axesBottomLeft =
          (axesDimensions && axesDimensions.bottom && sumObjBy(axesDimensions.bottom, 'left')) || 0
        return paddingLeft + Math.max(axesLeftWidth, axesTopLeft, axesBottomLeft)
      }
    ),

  gridY: () =>
    Memoize(
      state => [(state && state.padding && state.padding.top) || 0, state && state.axesDimensions],
      (paddingTop, axesDimensions) => {
        const axesTopHeight =
          (axesDimensions && axesDimensions.top && sumObjBy(axesDimensions.top, 'height')) || 0
        const axesLeftTop =
          (axesDimensions && axesDimensions.left && sumObjBy(axesDimensions.left, 'top')) || 0
        const axesRightTop =
          (axesDimensions && axesDimensions.right && sumObjBy(axesDimensions.right, 'top')) || 0
        return paddingTop + Math.max(axesTopHeight, axesLeftTop, axesRightTop)
      }
    ),

  gridWidth: () =>
    Memoize(
      state => [state && state.width, state && state.padding, state && state.axisDimensions],
      (width, padding, axisDimensions) => {
        const paddingLeft = padding && padding.left
        const paddingRight = padding && padding.right
        const axesLeftWidth =
          (axisDimensions && axisDimensions.left && sumObjBy(axisDimensions.left, 'width')) || 0
        const axesRightWidth =
          (axisDimensions && axisDimensions.right && sumObjBy(axisDimensions.right, 'width')) || 0
        const axesTopLeft =
          (axisDimensions && axisDimensions.top && sumObjBy(axisDimensions.top, 'left')) || 0
        const axesTopRight =
          (axisDimensions && axisDimensions.top && sumObjBy(axisDimensions.top, 'right')) || 0
        const axesBottomLeft =
          (axisDimensions && axisDimensions.bottom && sumObjBy(axisDimensions.bottom, 'left')) || 0
        const axesBottomRight =
          (axisDimensions && axisDimensions.bottom && sumObjBy(axisDimensions.bottom, 'right')) || 0

        return (
          (width || 0) -
          (paddingLeft || 0) -
          (paddingRight || 0) -
          Math.max(axesLeftWidth, axesTopLeft, axesBottomLeft) -
          Math.max(axesRightWidth, axesTopRight, axesBottomRight)
        )
      }
    ),

  gridHeight: () =>
    Memoize(
      state => [state && state.height, state && state.padding, state && state.axisDimensions],
      (height, padding, axisDimensions) => {
        const paddingTop = padding && padding.top
        const paddingBottom = padding && padding.right
        const axesTopHeight =
          (axisDimensions && axisDimensions.top && sumObjBy(axisDimensions.top, 'height')) || 0
        const axesBottomHeight =
          (axisDimensions && axisDimensions.bottom && sumObjBy(axisDimensions.bottom, 'height')) ||
          0
        const axesLeftTop =
          (axisDimensions && axisDimensions.left && sumObjBy(axisDimensions.left, 'top')) || 0
        const axesLeftBottom =
          (axisDimensions && axisDimensions.left && sumObjBy(axisDimensions.left, 'bottom')) || 0
        const axesRightTop =
          (axisDimensions && axisDimensions.right && sumObjBy(axisDimensions.right, 'top')) || 0
        const axesRightBottom =
          (axisDimensions && axisDimensions.right && sumObjBy(axisDimensions.right, 'bottom')) || 0

        return (
          (height || 0) -
          (paddingTop || 0) -
          (paddingBottom || 0) -
          Math.max(axesTopHeight, axesLeftTop, axesRightTop) -
          Math.max(axesBottomHeight, axesLeftBottom, axesRightBottom)
        )
      }
    ),
}

function sumObjBy (obj, str) {
  return Object.keys(obj)
    .map(key => obj[key])
    .reduce((prev, curr) => prev + curr[str] || 0, 0)
}
